### DISCUSSION TITLE
MUI 5 Newest DesktopDatePicker validation not triggered

### DISCUSSION THREAD

**Discussion by keemor** [react-hook-form/react-hook-form#8728] [Q&A] [answered]
Hi :)

Validation works in this older example:
https://codesandbox.io/s/material-demo-forked-hvgxyn?file=/demo.tsx

![image](https://user-images.githubusercontent.com/1793635/180761710-499bfe2f-32f7-4700-af13-3642e350beef.png)

I created similiar solution based on   "@mui/material": "latest" & "@mui/x-date-pickers": "latest",

https://codesandbox.io/s/react-hook-form-date-picker-6d8x6d?file=/src/form/DateField.tsx

It works for require, but validation does not work and I can't figure out why

![image](https://user-images.githubusercontent.com/1793635/180762886-dd790f13-62c1-4643-b765-0acd1d4f404c.png)

Could you help me pls?

**Comment by Moshyfawn** [comment, marked as answer]
Can you provide the info on what's the expected behavior? What validation?

The only validation rule you're passing is the `required` rule:
```jsx
<Controller
  rules={{
    required: required ? "Required" : ""
  }}
/>
```

**Reply by keemor** [reply]
I would like to see Invalid Date Format message as it works here:

https://codesandbox.io/s/material-demo-forked-hvgxyn?file=/demo.tsx

In above example there are no rules, but it still shows error Invalid Date Format for 99/99/9999

**Reply by Moshyfawn** [reply]
It's the MUI `DatePicker` specific behavior and has nothing to do with RHF; RHF doesn't create any custom format error, but rather simply compares your validation rules to the input value and throws user-defined errors

**Comment by keemor** [comment]
I added custom validate method and seems to work, is this approach ok?

```js
rules={{
  required: required ? "Required" : "",
  validate: () => (!value || isValid(value) ? true : "Wrong format")
}}
```

**Reply by Moshyfawn** [reply]
Yep, fair!

For convenience, you can split your validate callback return into multiple validation fns:
``` js
validate: {
  short: (v) => v.length > 5 || "Too short",
  long: (v) => v.length < 10 || "Too long"
}
```

**Comment by keemor** [comment]
https://codesandbox.io/s/react-hook-form-date-picker-6d8x6d?file=/src/form/DateField.tsx

I noticed one more problem:

1. Enter invalid date 88/88/8888 and hit submit - error is shown
2. Open calendar and pick date from calendar - value is ok, but error stays - it should disappear

![image](https://user-images.githubusercontent.com/1793635/180955796-d9a9efd2-3b3d-418d-a4b5-b3fec6e45924.png)

Any hint on this one?

If I clear second field (required one) and hit submit "require" message is shown, but after selecting value from calendar error message disappears as expected

**Reply by Moshyfawn** [reply]
Your validation rule is subscribed to the field's value via `useWatch` instead of using the newest input value from the `validate` callback. Using `Controller`, you isolate the re-renders / components updates to that specific component wrapped in `Controller`, hence the field value used from `useWatch` is stale "inside" the validate callback.

You can notice that the validation error value is actually 1 render behind: if you change the input's value again, you'll see that it's the previous validation error state that's shown

Basically, **instead of**
```jsx
const date = useWatch()

return <Controller rules={{ validate: () => validateDate(date) }} />
```

**do**
```jsx
<Controller rules={{ validate: date => validateDate(date) }} />
```

**Reply by keemor** [reply]
It works, thank you very much for your time, I totally missed that. I updated CSB to work as expected! :)

**Reply by Moshyfawn** [reply]
No worries! I'm glad I could help!

**Reply by bluebill1049** [reply]
Thanks very much @Moshyfawn for the help ❤️
